### PR TITLE
Try to fix callbacks

### PR DESCRIPTION
**Pull request by kanav99**
The new `find_bracket` feature doesn't work either, in fact it brings an unexpected issue. I think we should think of a new approach to fix this floating point issue.

1. In case of repeated event, currently we check using derivative. If value of condition is small enough and is increasing, we say that sign is +ve on the left hand (similar for decreasing ). But when we find the callback time, we use actual value of condition at some point ahead until it switches sign. I think this is something wrong, for both detection and time calculation we should do the same. Otherwise such errors are bound to happen.

2. We also do need a `OnceContinuousCallback` that only runs once. or maybe a good way to switch off the callback. A part of code can be simplified if we make a way to switch off the callback which is documented.

3. I support complete refactoring of the callback code as this issue might be something else.

I have some private code that had been shared to me by users. Unless all of that is working, this isnt fixed

@ChrisRackauckas @YingboMa @devmotion do you have any thoughts?

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
>I think this is something wrong, for both detection and time calculation we should do the same.

Why? What's a better scheme?

**Comment by ChrisRackauckas** [comment]
>or maybe a good way to switch off the callback. A part of code can be simplified if we make a way to switch off the callback which is documented.

A good way to switch on and off callbacks is probably the answer.

**Comment by kanav99** [comment]
Okay so I wrote a custom `bisection` (very rough needs refining) but I realized two things:
1. Using `Θs` instead of `ts` was a bad choice because if we change `dt` later on, an if that's too big or too smaller than last step, then `abst = integrator.tprev + integrator.dt*Θ` could do funny stuff. The MWE I had made `dt` too small so basically `abst` completely same during the 12 loop. 
2. Certainly Roots.jl still have some bugs. Changing the `bisection` to `find_zero(...., BisectionExact())` made `event_detection_tests.jl` error.

Its late night here and would continue making it better in the morning.

**Comment by kanav99** [comment]
Running OrdinaryDiffEq test locally. @ChrisRackauckas please check the `bisection` function

File: src/callbacks.jl [hunk 1]
```diff
@@ -560,7 +560,37 @@ end
...
+
+# rough implementation, needs multiple type handling
+# always ensures that if r = bisection(f, (x0, x1))
+# then either f(nextfloat(r)) == 0 or f(nextfloat(r)) * f(r) < 0
+function bisection(f, tup, tdir)
+  x0, x1 = tup
+  @assert f(x0) * f(x1) < 0.0
+  left = x0
+  right = x1
+  prevfloat_tdir = (tdir === 1.0 ? prevfloat : nextfloat)
```
Comment (devmotion) [inline src/callbacks.jl]: What's the reason for not writing
```suggestion
  prevfloat_tdir = isone(tdir) ? prevfloat : nextfloat
```

File: src/callbacks.jl [hunk 2]
```diff
@@ -560,7 +560,37 @@ end
...
+  while true
+    @assert f(left) * f(right) < 0.0
+    mid = (left + right) / 2
+    y = f(mid)
+    if y === 0.0
```
Comment (devmotion) [inline src/callbacks.jl]: Same here, why not
```suggestion
    if iszero(y)
```
?

File: src/callbacks.jl [hunk 3]
```diff
@@ -560,7 +560,37 @@ end
...
+      # we are in the region of zero
+      mid = prevfloat_tdir(mid)
+      while f(mid) === 0.0
```
Comment (devmotion) [inline src/callbacks.jl]: And here?
```suggestion
      while iszero(f(mid))
```

File: src/callbacks.jl (hunk 1 above)
Comment (kanav99) [reply src/callbacks.jl]: Oh yes, AD test would fail otherwise, my bad

Comment (devmotion) [reply src/callbacks.jl]: The `@assert` statements could be problematic as well, I guess, but I assumed these are mainly for debugging right now?

**Comment by kanav99** [comment]
AD tests still fail after the change

**Comment by kanav99** [comment]
Any help why these AD test are failing? I plugged `@show typeof(...)` in many places inside `bisection`
```julia
function bisection(f, tup, tdir)
  x0, x1 = tup
  @show typeof(x0)
  @show typeof(x1)
  @assert f(x0) * f(x1) < 0.0
  left = x0
  right = x1
  prevfloat_tdir = isone(tdir) ? prevfloat : nextfloat
  while true
    @assert f(left) * f(right) < 0.0
    mid = (left + right) / 2
    @show typeof(mid)
    y = f(mid)
    @show typeof(y)
    if iszero(y)
      # we are in the region of zero
      mid = prevfloat_tdir(mid)
      @show typeof(mid)
      while iszero(f(mid))
        mid = prevfloat_tdir(mid)
        @show typeof(mid)
      end
      return mid
    end
    (left === mid || right === mid #= i dont think second condition would happen =#) && return left
    if sign(y) === sign(f(left))
      left = mid
      @show typeof(left)
    else
      right = mid
      @show typeof(right)
    end
  end
end
```
All of them are duals which means the types are getting inserted well. 

Test code:
```julia
using Test
using OrdinaryDiffEq, Calculus, ForwardDiff

function f(du,u,p,t)
  du[1] = -p[1]
  du[2] = p[2]
end

for x in 0:0.001:5
  called = false
  function test_f(p)
    cb = ContinuousCallback((u,t,i) -> u[1], (integrator)->(called=true;integrator.p[2]=zero(integrator.p[2])))
    prob = ODEProblem(f,eltype(p).([1.0,0.0]),eltype(p).((0.0,1.0)),copy(p))
    integrator = init(prob,Tsit5(),abstol=1e-14,reltol=1e-14,callback=cb)
    step!(integrator)
    solve!(integrator).u[end]
  end
  p = [2.0, x]
  called = false
  findiff = Calculus.finite_difference_jacobian(test_f,p)
  @test called
  called = false
  fordiff = ForwardDiff.jacobian(test_f,p)
  @test called
  @test findiff ≈ fordiff
end
```

**Comment by kanav99** [comment]
I realized its the problem due to branching inside the Bisection Methods. This fails even on `Roots.Bisection`. I think we should make a branch for AD and Non-AD cases.

**Comment by ChrisRackauckas** [comment]
It's sad we can't use a safe one with AD. Is there any way to stabilize a false position method?

**Comment by ChrisRackauckas** [comment]
Actually let's go with this for now, and then in the future change it so that we take over the dual dispatch and utilize the pushforward derivative, which can be directly applied at the solution. The adjoint version can be found in https://math.mit.edu/~stevenj/18.336/adjoint.pdf . We can do something similar (it's just the implicit function theorem).

**Comment by kanav99** [comment]
> It's sad we can't use a safe one with AD. Is there any way to stabilize a false position method?

Problem with AlefeldPotraShi right now is that it throws convergence error when I set tolerance zero, otherwise there is no problem. I will wait for it to be stable.

 > Actually let's go with this for now, and then in the future change it so that we take over the dual dispatch and utilize the pushforward derivative, which can be directly applied at the solution. The adjoint version can be found in https://math.mit.edu/~stevenj/18.336/adjoint.pdf . We can do something similar (it's just the implicit function theorem).

Sounds like a plan. I am running OrdinaryDiffEq test. Most of the important one's have passed. All the new MWEs passed. @HelgavonLichtenstein

File: src/callbacks.jl (hunk 1 above)
Comment (kanav99) [reply src/callbacks.jl]: I can change them to proper errors. But those are very important assertions, in case of any edge case, we need a solid starting point to debug

**Comment by kanav99** [comment]
Test pass @ChrisRackauckas , should I add some tests for the new bisection function?

**Comment by ChrisRackauckas** [comment]
Yes, just add a few tests so we trust the bisection method, and let's see how this goes.

**Comment by kanav99** [comment]
OrdinaryDiffEq tests running here - https://colab.research.google.com/drive/1okm12KhJIyEhfQYHj-MbDykv0T_rLHXM

**Comment by kanav99** [comment]
All integrator and interface tests pass. Running convergence tests, but they don't have any callbacks. Can you look at the latest commit @ChrisRackauckas @devmotion for any errors?

**Comment by ChrisRackauckas** [comment]
The last test is fine. This looks good to merge then. You happy with it?

**Comment by kanav99** [comment]
Done. It looks good to me to be merged.

File: src/callbacks.jl [hunk 4]
```diff
@@ -568,15 +568,16 @@ end
 # then either f(nextfloat(r)) == 0 or f(nextfloat(r)) * f(r) < 0
 function bisection(f, tup, tdir; maxiters=100)
   x0, x1 = tup
-  (f(x0) * f(x1) >= 0.0) && error("Non bracketing interval passed in bisection method. Please report the error in DiffEqBase.")
+  fx0x1 = f(x0) * f(x1)
+  (fx0x1 >= zero(fx0x1)) && error("Non bracketing interval passed in bisection method. Please report the error in DiffEqBase.")
```
Comment (devmotion) [inline src/callbacks.jl]: Maybe just compute `zero(fx0x1)` once in the beginning instead of recomputing it in the `while`? Although I guess it shouldn't have a severe impact on performance, might be compiled away?

Comment (kanav99) [reply src/callbacks.jl]: I hope that zeros are easily compiled. Still let me do that.

Comment (YingboMa) [reply src/callbacks.jl]: Zero isn't free to compute. Doing manual loop invariant code motion is better.

**Comment by kanav99** [comment]
One more thing, I have set maxiters to 100 right now. iters bisection takes is proportional to the log of time interval. For `x1 - x0 = 25.967830436978204`, it takes 50 iters. So a limit of `maxiters = 100` would be fine?

**Comment by YingboMa** [comment]
Do we have to set `maxiters`? Stiff solvers could potentially step for a very long interval.

**Comment by ChrisRackauckas** [comment]
I don't think we need to set maxiters. 2^100 = 1.2676506e+30, so the default is essentially going until floating point error, unless it doesn't on a problem where people want really high precision, in which case having maxiters doesn't make sense because they want high precision.

**Comment by kanav99** [comment]
I want to set it for now, I have made sure this loop wont be stuck in an infinite loop, but just for safety. I can increase it, but I think we should set it for now, just to prevent infinite loops.

**Comment by devmotion** [comment]
>  I think we should set it for now, just to prevent infinite loops.

Couldn't you check something else like that you actually shrink the bracketing interval to make sure it does terminate at some point? Something that's independent of the number of iteration steps?

**Comment by kanav99** [comment]
I have already added this line - https://github.com/JuliaDiffEq/DiffEqBase.jl/pull/456/files#diff-f2ec50e45e2c0d098a69b4e7c72f86f3R591 which checks if mid is one of the endpoints of bracket, that should probably always check for infinite loops. If you think thats enough, I will remove the maxiters.

**Comment by kanav99** [comment]
Some good news from JuliaMath/Roots.jl#174 too

**Comment by ChrisRackauckas** [comment]
yeah, that's really good news and probably the final solution, but let's at least get this through for now.
